embodiedscan/visualization/utils.py

```diff
@@ -39,23 +40,104 @@ def _box_add_thickness(box, thickness):
     return results
 
 
-def _9dof_to_box(box, label, color_selector):
+def _9dof_to_box(box, label=None, color_selector=None, color=None):
+    """Convert 9-DoF box from array/tensor to open3d.OrientedBoundingBox.
+
+    Args:
+        box (numpy.ndarray or torch.Tensor): 9-DoF box with shape (9,).
+        label (int, optional): Label of the box. Defaults to None.
+        color_selector (:obj:`ColorSelector`): Color selector for boxes.
+            Defaults to None.
+        color (tuple[int]): Color of the box.
+            You can directly specify the color.
+            If you do, the color_selector and label will be ignored.
+            Defaults to None.
+    """
     if isinstance(box, list):
         box = np.array(box)
+    if isinstance(box, Tensor):
+        box = box.cpu().numpy()
     center = box[:3].reshape(3, 1)
     scale = box[3:6].reshape(3, 1)
     rot = box[6:].reshape(3, 1)
     rot_mat = o3d.geometry.OrientedBoundingBox.get_rotation_matrix_from_zxy(
         rot)
     geo = o3d.geometry.OrientedBoundingBox(center, rot_mat, scale)
 
-    color = color_selector.get_color(label)
-    color = [x / 255.0 for x in color]
-    geo.color = color
+    if color is not None:
+        geo.color = [x / 255.0 for x in color]
+        return geo
+
+    if label is not None and color_selector is not None:
+        color = color_selector.get_color(label)
+        color = [x / 255.0 for x in color]
+        geo.color = color
     return geo
 
 
+def nms_filter(pred_results, iou_thr=0.15, score_thr=0.075, topk_per_class=10):
+    """Non-Maximum Suppression for 3D Euler boxes. Additionally, only the top-k
+    boxes will be kept for each category to avoid redundant boxes in the
+    visualization.
+
+    Args:
+        pred_results (:obj:`InstanceData`):
+            Results predicted by the model
+        iou_thr (float): IoU thresholds for NMS. Defaults to 0.15.
+        score_thr (float): Score thresholds.
+            Instances with scores below thresholds will not be kept.
+            Defaults to 0.075.
+        topk_per_class (int): Number of instances kept per category.
+
+    Returns:
```

Comment:
numpy.ndarray[float], np.ndarray[int]: Filtered boxes with shape (N, 9) and labels with shape (N,).